// playwright@1.56.0-alpha-1756505518000/lib/mcp/test/listTests.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var listTests_exports = {};
__export(listTests_exports, {
  listTests: () => listTests
});
module.exports = __toCommonJS(listTests_exports);
var import_path = __toESM(require("path"));
var import_bundle = require("../sdk/bundle");
var import_tool = require("./tool.js");
const listTests = (0, import_tool.defineTool)({
  schema: {
    name: "playwright_test_list_tests",
    title: "List tests",
    description: "List tests",
    inputSchema: import_bundle.z.object({}),
    type: "readOnly"
  },
  handle: async (context, params) => {
    const reporter = new ListModeReporter();
    const testRunner = await context.createTestRunner();
    await testRunner.listTests(reporter, {});
    if (reporter.hasErrors())
      throw new Error(reporter.content());
    return {
      content: [{ type: "text", text: reporter.content() }]
    };
  }
});
class ListModeReporter {
  constructor() {
    this._lines = [];
    this._hasErrors = false;
  }
  onBegin(config, suite) {
    this._lines.push(`Listing tests:`);
    const tests = suite.allTests();
    const files = /* @__PURE__ */ new Set();
    for (const test of tests) {
      const [, projectName, , ...titles] = test.titlePath();
      const location = `${import_path.default.relative(config.rootDir, test.location.file)}:${test.location.line}:${test.location.column}`;
      const projectTitle = projectName ? `[${projectName}] \u203A ` : "";
      this._lines.push(`  [id=${test.id}] ${projectTitle}${location} \u203A ${titles.join(" \u203A ")}`);
      files.add(test.location.file);
    }
    this._lines.push(`Total: ${tests.length} ${tests.length === 1 ? "test" : "tests"} in ${files.size} ${files.size === 1 ? "file" : "files"}`);
  }
  onError(error) {
    this._hasErrors = true;
    this._lines.push(error.stack || error.message || error.value || "");
  }
  hasErrors() {
    return this._hasErrors;
  }
  content() {
    return this._lines.join("\n");
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  listTests
});
