// playwright@1.56.0-alpha-1756505518000/lib/mcp/test/runTests.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var runTests_exports = {};
__export(runTests_exports, {
  runTests: () => runTests
});
module.exports = __toCommonJS(runTests_exports);
var import_utils = require("playwright-core/lib/utils");
var import_bundle = require("../sdk/bundle");
var import_base = require("../../reporters/base");
var import_list = __toESM(require("../../reporters/list"));
var import_tool = require("./tool");
var import_streams = require("./streams");
const runTests = (0, import_tool.defineTool)({
  schema: {
    name: "playwright_test_run_tests",
    title: "Run tests",
    description: "Run tests",
    inputSchema: import_bundle.z.object({
      tests: import_bundle.z.array(import_bundle.z.object({
        id: import_bundle.z.string().describe("Test ID to run."),
        title: import_bundle.z.string().describe("Human readable test title for granting permission to run the test.")
      })).optional().describe("Tests to run. All tests are run if not provided.")
    }),
    type: "readOnly"
  },
  handle: async (context, params) => {
    const stream = new import_streams.StringWriteStream();
    const screen = {
      ...import_base.terminalScreen,
      isTTY: false,
      colors: import_utils.noColors,
      stdout: stream,
      stderr: stream
    };
    const configDir = context.configLocation.configDir;
    const reporter = new import_list.default({ configDir, screen });
    const testRunner = await context.createTestRunner();
    const result = await testRunner.runTests(reporter, {
      testIds: params.tests?.map((test) => test.id),
      // For automatic recovery
      timeout: 0,
      workers: 1
    });
    const text = stream.content();
    return {
      content: [
        { type: "text", text }
      ],
      isError: result.status !== "passed"
    };
  }
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  runTests
});
