// @metagptx/web-sdk@0.0.68/dist/plugins.js
import*as e from"@babel/parser";import*as t from"@babel/parser";import n from"@babel/traverse";import*as r from"@babel/types";import*as i from"@babel/types";import{transformWithEsbuild as a}from"vite";import o from"node:fs";import s from"node:path";var c=44,l=59,u=`ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/`,d=new Uint8Array(64),f=new Uint8Array(128);for(let e=0;e<u.length;e++){let t=u.charCodeAt(e);d[e]=t,f[t]=e}function p(e,t,n){let r=t-n;r=r<0?-r<<1|1:r<<1;do{let t=r&31;r>>>=5,r>0&&(t|=32),e.write(d[t])}while(r>0);return t}var m=1024*16,h=typeof TextDecoder<`u`?new TextDecoder:typeof Buffer<`u`?{decode(e){let t=Buffer.from(e.buffer,e.byteOffset,e.byteLength);return t.toString()}}:{decode(e){let t=``;for(let n=0;n<e.length;n++)t+=String.fromCharCode(e[n]);return t}},g=class{constructor(){this.pos=0,this.out=``,this.buffer=new Uint8Array(m)}write(e){let{buffer:t}=this;t[this.pos++]=e,this.pos===m&&(this.out+=h.decode(t),this.pos=0)}flush(){let{buffer:e,out:t,pos:n}=this;return n>0?t+h.decode(e.subarray(0,n)):t}};function _(e){let t=new g,n=0,r=0,i=0,a=0;for(let o=0;o<e.length;o++){let s=e[o];if(o>0&&t.write(l),s.length===0)continue;let u=0;for(let e=0;e<s.length;e++){let o=s[e];e>0&&t.write(c),u=p(t,o[0],u),o.length!==1&&(n=p(t,o[1],n),r=p(t,o[2],r),i=p(t,o[3],i),o.length!==4&&(a=p(t,o[4],a)))}}return t.flush()}var v=class e{constructor(t){this.bits=t instanceof e?t.bits.slice():[]}add(e){this.bits[e>>5]|=1<<(e&31)}has(e){return!!(this.bits[e>>5]&1<<(e&31))}},y=class e{constructor(e,t,n){this.start=e,this.end=t,this.original=n,this.intro=``,this.outro=``,this.content=n,this.storeName=!1,this.edited=!1,this.previous=null,this.next=null}appendLeft(e){this.outro+=e}appendRight(e){this.intro+=e}clone(){let t=new e(this.start,this.end,this.original);return t.intro=this.intro,t.outro=this.outro,t.content=this.content,t.storeName=this.storeName,t.edited=this.edited,t}contains(e){return this.start<e&&e<this.end}eachNext(e){let t=this;for(;t;)e(t),t=t.next}eachPrevious(e){let t=this;for(;t;)e(t),t=t.previous}edit(e,t,n){return this.content=e,n||(this.intro=``,this.outro=``),this.storeName=t,this.edited=!0,this}prependLeft(e){this.outro=e+this.outro}prependRight(e){this.intro=e+this.intro}reset(){this.intro=``,this.outro=``,this.edited&&(this.content=this.original,this.storeName=!1,this.edited=!1)}split(t){let n=t-this.start,r=this.original.slice(0,n),i=this.original.slice(n);this.original=r;let a=new e(t,this.end,i);return a.outro=this.outro,this.outro=``,this.end=t,this.edited?(a.edit(``,!1),this.content=``):this.content=r,a.next=this.next,a.next&&(a.next.previous=a),a.previous=this,this.next=a,a}toString(){return this.intro+this.content+this.outro}trimEnd(e){if(this.outro=this.outro.replace(e,``),this.outro.length)return!0;let t=this.content.replace(e,``);if(t.length)return t!==this.content&&(this.split(this.start+t.length).edit(``,void 0,!0),this.edited&&this.edit(t,this.storeName,!0)),!0;if(this.edit(``,void 0,!0),this.intro=this.intro.replace(e,``),this.intro.length)return!0}trimStart(e){if(this.intro=this.intro.replace(e,``),this.intro.length)return!0;let t=this.content.replace(e,``);if(t.length){if(t!==this.content){let e=this.split(this.end-t.length);this.edited&&e.edit(t,this.storeName,!0),this.edit(``,void 0,!0)}return!0}else if(this.edit(``,void 0,!0),this.outro=this.outro.replace(e,``),this.outro.length)return!0}};function b(){return typeof globalThis<`u`&&typeof globalThis.btoa==`function`?e=>globalThis.btoa(unescape(encodeURIComponent(e))):typeof Buffer==`function`?e=>Buffer.from(e,`utf-8`).toString(`base64`):()=>{throw Error("Unsupported environment: `window.btoa` or `Buffer` should be supported.")}}const x=b();var S=class{constructor(e){this.version=3,this.file=e.file,this.sources=e.sources,this.sourcesContent=e.sourcesContent,this.names=e.names,this.mappings=_(e.mappings),e.x_google_ignoreList!==void 0&&(this.x_google_ignoreList=e.x_google_ignoreList),e.debugId!==void 0&&(this.debugId=e.debugId)}toString(){return JSON.stringify(this)}toUrl(){return`data:application/json;charset=utf-8;base64,`+x(this.toString())}};function C(e){let t=e.split(`
`),n=t.filter(e=>/^\t+/.test(e)),r=t.filter(e=>/^ {2,}/.test(e));if(n.length===0&&r.length===0)return null;if(n.length>=r.length)return`	`;let i=r.reduce((e,t)=>{let n=/^ +/.exec(t)[0].length;return Math.min(n,e)},1/0);return Array(i+1).join(` `)}function w(e,t){let n=e.split(/[/\\]/),r=t.split(/[/\\]/);for(n.pop();n[0]===r[0];)n.shift(),r.shift();if(n.length){let e=n.length;for(;e--;)n[e]=`..`}return n.concat(r).join(`/`)}const T=Object.prototype.toString;function E(e){return T.call(e)===`[object Object]`}function D(e){let t=e.split(`
`),n=[];for(let e=0,r=0;e<t.length;e++)n.push(r),r+=t[e].length+1;return function(e){let t=0,r=n.length;for(;t<r;){let i=t+r>>1;e<n[i]?r=i:t=i+1}let i=t-1,a=e-n[i];return{line:i,column:a}}}const O=/\w/;var k=class{constructor(e){this.hires=e,this.generatedCodeLine=0,this.generatedCodeColumn=0,this.raw=[],this.rawSegments=this.raw[this.generatedCodeLine]=[],this.pending=null}addEdit(e,t,n,r){if(t.length){let i=t.length-1,a=t.indexOf(`
`,0),o=-1;for(;a>=0&&i>a;){let i=[this.generatedCodeColumn,e,n.line,n.column];r>=0&&i.push(r),this.rawSegments.push(i),this.generatedCodeLine+=1,this.raw[this.generatedCodeLine]=this.rawSegments=[],this.generatedCodeColumn=0,o=a,a=t.indexOf(`
`,a+1)}let s=[this.generatedCodeColumn,e,n.line,n.column];r>=0&&s.push(r),this.rawSegments.push(s),this.advance(t.slice(o+1))}else this.pending&&(this.rawSegments.push(this.pending),this.advance(t));this.pending=null}addUneditedChunk(e,t,n,r,i){let a=t.start,o=!0,s=!1;for(;a<t.end;){if(n[a]===`
`)r.line+=1,r.column=0,this.generatedCodeLine+=1,this.raw[this.generatedCodeLine]=this.rawSegments=[],this.generatedCodeColumn=0,o=!0,s=!1;else{if(this.hires||o||i.has(a)){let t=[this.generatedCodeColumn,e,r.line,r.column];this.hires===`boundary`?O.test(n[a])?s||(this.rawSegments.push(t),s=!0):(this.rawSegments.push(t),s=!1):this.rawSegments.push(t)}r.column+=1,this.generatedCodeColumn+=1,o=!1}a+=1}this.pending=null}advance(e){if(!e)return;let t=e.split(`
`);if(t.length>1){for(let e=0;e<t.length-1;e++)this.generatedCodeLine++,this.raw[this.generatedCodeLine]=this.rawSegments=[];this.generatedCodeColumn=0}this.generatedCodeColumn+=t[t.length-1].length}};const A=`
`,j={insertLeft:!1,insertRight:!1,storeName:!1};var M=class e{constructor(e,t={}){let n=new y(0,e.length,e);Object.defineProperties(this,{original:{writable:!0,value:e},outro:{writable:!0,value:``},intro:{writable:!0,value:``},firstChunk:{writable:!0,value:n},lastChunk:{writable:!0,value:n},lastSearchedChunk:{writable:!0,value:n},byStart:{writable:!0,value:{}},byEnd:{writable:!0,value:{}},filename:{writable:!0,value:t.filename},indentExclusionRanges:{writable:!0,value:t.indentExclusionRanges},sourcemapLocations:{writable:!0,value:new v},storedNames:{writable:!0,value:{}},indentStr:{writable:!0,value:void 0},ignoreList:{writable:!0,value:t.ignoreList},offset:{writable:!0,value:t.offset||0}}),this.byStart[0]=n,this.byEnd[e.length]=n}addSourcemapLocation(e){this.sourcemapLocations.add(e)}append(e){if(typeof e!=`string`)throw TypeError(`outro content must be a string`);return this.outro+=e,this}appendLeft(e,t){if(e+=this.offset,typeof t!=`string`)throw TypeError(`inserted content must be a string`);this._split(e);let n=this.byEnd[e];return n?n.appendLeft(t):this.intro+=t,this}appendRight(e,t){if(e+=this.offset,typeof t!=`string`)throw TypeError(`inserted content must be a string`);this._split(e);let n=this.byStart[e];return n?n.appendRight(t):this.outro+=t,this}clone(){let t=new e(this.original,{filename:this.filename,offset:this.offset}),n=this.firstChunk,r=t.firstChunk=t.lastSearchedChunk=n.clone();for(;n;){t.byStart[r.start]=r,t.byEnd[r.end]=r;let e=n.next,i=e&&e.clone();i&&(r.next=i,i.previous=r,r=i),n=e}return t.lastChunk=r,this.indentExclusionRanges&&(t.indentExclusionRanges=this.indentExclusionRanges.slice()),t.sourcemapLocations=new v(this.sourcemapLocations),t.intro=this.intro,t.outro=this.outro,t}generateDecodedMap(e){e||={};let t=0,n=Object.keys(this.storedNames),r=new k(e.hires),i=D(this.original);return this.intro&&r.advance(this.intro),this.firstChunk.eachNext(e=>{let a=i(e.start);e.intro.length&&r.advance(e.intro),e.edited?r.addEdit(t,e.content,a,e.storeName?n.indexOf(e.original):-1):r.addUneditedChunk(t,e,this.original,a,this.sourcemapLocations),e.outro.length&&r.advance(e.outro)}),this.outro&&r.advance(this.outro),{file:e.file?e.file.split(/[/\\]/).pop():void 0,sources:[e.source?w(e.file||``,e.source):e.file||``],sourcesContent:e.includeContent?[this.original]:void 0,names:n,mappings:r.raw,x_google_ignoreList:this.ignoreList?[t]:void 0}}generateMap(e){return new S(this.generateDecodedMap(e))}_ensureindentStr(){this.indentStr===void 0&&(this.indentStr=C(this.original))}_getRawIndentString(){return this._ensureindentStr(),this.indentStr}getIndentString(){return this._ensureindentStr(),this.indentStr===null?`	`:this.indentStr}indent(e,t){let n=/^[^\r\n]/gm;if(E(e)&&(t=e,e=void 0),e===void 0&&(this._ensureindentStr(),e=this.indentStr||`	`),e===``)return this;t||={};let r={};if(t.exclude){let e=typeof t.exclude[0]==`number`?[t.exclude]:t.exclude;e.forEach(e=>{for(let t=e[0];t<e[1];t+=1)r[t]=!0})}let i=t.indentStart!==!1,a=t=>i?`${e}${t}`:(i=!0,t);this.intro=this.intro.replace(n,a);let o=0,s=this.firstChunk;for(;s;){let t=s.end;if(s.edited)r[o]||(s.content=s.content.replace(n,a),s.content.length&&(i=s.content[s.content.length-1]===`
`));else for(o=s.start;o<t;){if(!r[o]){let t=this.original[o];t===`
`?i=!0:t!==`\r`&&i&&(i=!1,o===s.start?s.prependRight(e):(this._splitChunk(s,o),s=s.next,s.prependRight(e)))}o+=1}o=s.end,s=s.next}return this.outro=this.outro.replace(n,a),this}insert(){throw Error(`magicString.insert(...) is deprecated. Use prependRight(...) or appendLeft(...)`)}insertLeft(e,t){return j.insertLeft||(console.warn(`magicString.insertLeft(...) is deprecated. Use magicString.appendLeft(...) instead`),j.insertLeft=!0),this.appendLeft(e,t)}insertRight(e,t){return j.insertRight||(console.warn(`magicString.insertRight(...) is deprecated. Use magicString.prependRight(...) instead`),j.insertRight=!0),this.prependRight(e,t)}move(e,t,n){if(e+=this.offset,t+=this.offset,n+=this.offset,n>=e&&n<=t)throw Error(`Cannot move a selection inside itself`);this._split(e),this._split(t),this._split(n);let r=this.byStart[e],i=this.byEnd[t],a=r.previous,o=i.next,s=this.byStart[n];if(!s&&i===this.lastChunk)return this;let c=s?s.previous:this.lastChunk;return a&&(a.next=o),o&&(o.previous=a),c&&(c.next=r),s&&(s.previous=i),r.previous||(this.firstChunk=i.next),i.next||(this.lastChunk=r.previous,this.lastChunk.next=null),r.previous=c,i.next=s||null,c||(this.firstChunk=r),s||(this.lastChunk=i),this}overwrite(e,t,n,r){return r||={},this.update(e,t,n,{...r,overwrite:!r.contentOnly})}update(e,t,n,r){if(e+=this.offset,t+=this.offset,typeof n!=`string`)throw TypeError(`replacement content must be a string`);if(this.original.length!==0){for(;e<0;)e+=this.original.length;for(;t<0;)t+=this.original.length}if(t>this.original.length)throw Error(`end is out of bounds`);if(e===t)throw Error(`Cannot overwrite a zero-length range – use appendLeft or prependRight instead`);this._split(e),this._split(t),r===!0&&(j.storeName||(console.warn(`The final argument to magicString.overwrite(...) should be an options object. See https://github.com/rich-harris/magic-string`),j.storeName=!0),r={storeName:!0});let i=r===void 0?!1:r.storeName,a=r===void 0?!1:r.overwrite;if(i){let n=this.original.slice(e,t);Object.defineProperty(this.storedNames,n,{writable:!0,value:!0,enumerable:!0})}let o=this.byStart[e],s=this.byEnd[t];if(o){let e=o;for(;e!==s;){if(e.next!==this.byStart[e.end])throw Error(`Cannot overwrite across a split point`);e=e.next,e.edit(``,!1)}o.edit(n,i,!a)}else{let r=new y(e,t,``).edit(n,i);s.next=r,r.previous=s}return this}prepend(e){if(typeof e!=`string`)throw TypeError(`outro content must be a string`);return this.intro=e+this.intro,this}prependLeft(e,t){if(e+=this.offset,typeof t!=`string`)throw TypeError(`inserted content must be a string`);this._split(e);let n=this.byEnd[e];return n?n.prependLeft(t):this.intro=t+this.intro,this}prependRight(e,t){if(e+=this.offset,typeof t!=`string`)throw TypeError(`inserted content must be a string`);this._split(e);let n=this.byStart[e];return n?n.prependRight(t):this.outro=t+this.outro,this}remove(e,t){if(e+=this.offset,t+=this.offset,this.original.length!==0){for(;e<0;)e+=this.original.length;for(;t<0;)t+=this.original.length}if(e===t)return this;if(e<0||t>this.original.length)throw Error(`Character is out of bounds`);if(e>t)throw Error(`end must be greater than start`);this._split(e),this._split(t);let n=this.byStart[e];for(;n;)n.intro=``,n.outro=``,n.edit(``),n=t>n.end?this.byStart[n.end]:null;return this}reset(e,t){if(e+=this.offset,t+=this.offset,this.original.length!==0){for(;e<0;)e+=this.original.length;for(;t<0;)t+=this.original.length}if(e===t)return this;if(e<0||t>this.original.length)throw Error(`Character is out of bounds`);if(e>t)throw Error(`end must be greater than start`);this._split(e),this._split(t);let n=this.byStart[e];for(;n;)n.reset(),n=t>n.end?this.byStart[n.end]:null;return this}lastChar(){if(this.outro.length)return this.outro[this.outro.length-1];let e=this.lastChunk;do{if(e.outro.length)return e.outro[e.outro.length-1];if(e.content.length)return e.content[e.content.length-1];if(e.intro.length)return e.intro[e.intro.length-1]}while(e=e.previous);return this.intro.length?this.intro[this.intro.length-1]:``}lastLine(){let e=this.outro.lastIndexOf(A);if(e!==-1)return this.outro.substr(e+1);let t=this.outro,n=this.lastChunk;do{if(n.outro.length>0){if(e=n.outro.lastIndexOf(A),e!==-1)return n.outro.substr(e+1)+t;t=n.outro+t}if(n.content.length>0){if(e=n.content.lastIndexOf(A),e!==-1)return n.content.substr(e+1)+t;t=n.content+t}if(n.intro.length>0){if(e=n.intro.lastIndexOf(A),e!==-1)return n.intro.substr(e+1)+t;t=n.intro+t}}while(n=n.previous);return e=this.intro.lastIndexOf(A),e===-1?this.intro+t:this.intro.substr(e+1)+t}slice(e=0,t=this.original.length-this.offset){if(e+=this.offset,t+=this.offset,this.original.length!==0){for(;e<0;)e+=this.original.length;for(;t<0;)t+=this.original.length}let n=``,r=this.firstChunk;for(;r&&(r.start>e||r.end<=e);){if(r.start<t&&r.end>=t)return n;r=r.next}if(r&&r.edited&&r.start!==e)throw Error(`Cannot use replaced character ${e} as slice start anchor.`);let i=r;for(;r;){r.intro&&(i!==r||r.start===e)&&(n+=r.intro);let a=r.start<t&&r.end>=t;if(a&&r.edited&&r.end!==t)throw Error(`Cannot use replaced character ${t} as slice end anchor.`);let o=i===r?e-r.start:0,s=a?r.content.length+t-r.end:r.content.length;if(n+=r.content.slice(o,s),r.outro&&(!a||r.end===t)&&(n+=r.outro),a)break;r=r.next}return n}snip(e,t){let n=this.clone();return n.remove(0,e),n.remove(t,n.original.length),n}_split(e){if(this.byStart[e]||this.byEnd[e])return;let t=this.lastSearchedChunk,n=t,r=e>t.end;for(;t;){if(t.contains(e))return this._splitChunk(t,e);if(t=r?this.byStart[t.end]:this.byEnd[t.start],t===n)return;n=t}}_splitChunk(e,t){if(e.edited&&e.content.length){let n=D(this.original)(t);throw Error(`Cannot split a chunk that has already been edited (${n.line}:${n.column} – "${e.original}")`)}let n=e.split(t);return this.byEnd[t]=e,this.byStart[t]=n,this.byEnd[n.end]=n,e===this.lastChunk&&(this.lastChunk=n),this.lastSearchedChunk=e,!0}toString(){let e=this.intro,t=this.firstChunk;for(;t;)e+=t.toString(),t=t.next;return e+this.outro}isEmpty(){let e=this.firstChunk;do if(e.intro.length&&e.intro.trim()||e.content.length&&e.content.trim()||e.outro.length&&e.outro.trim())return!1;while(e=e.next);return!0}length(){let e=this.firstChunk,t=0;do t+=e.intro.length+e.content.length+e.outro.length;while(e=e.next);return t}trimLines(){return this.trim(`[\\r\\n]`)}trim(e){return this.trimStart(e).trimEnd(e)}trimEndAborted(e){let t=RegExp((e||`\\s`)+`+$`);if(this.outro=this.outro.replace(t,``),this.outro.length)return!0;let n=this.lastChunk;do{let e=n.end,r=n.trimEnd(t);if(n.end!==e&&(this.lastChunk===n&&(this.lastChunk=n.next),this.byEnd[n.end]=n,this.byStart[n.next.start]=n.next,this.byEnd[n.next.end]=n.next),r)return!0;n=n.previous}while(n);return!1}trimEnd(e){return this.trimEndAborted(e),this}trimStartAborted(e){let t=RegExp(`^`+(e||`\\s`)+`+`);if(this.intro=this.intro.replace(t,``),this.intro.length)return!0;let n=this.firstChunk;do{let e=n.end,r=n.trimStart(t);if(n.end!==e&&(n===this.lastChunk&&(this.lastChunk=n.next),this.byEnd[n.end]=n,this.byStart[n.next.start]=n.next,this.byEnd[n.next.end]=n.next),r)return!0;n=n.next}while(n);return!1}trimStart(e){return this.trimStartAborted(e),this}hasChanged(){return this.original!==this.toString()}_replaceRegexp(e,t){function n(e,n){return typeof t==`string`?t.replace(/\$(\$|&|\d+)/g,(t,n)=>{if(n===`$`)return`$`;if(n===`&`)return e[0];let r=+n;return r<e.length?e[+n]:`$${n}`}):t(...e,e.index,n,e.groups)}function r(e,t){let n,r=[];for(;n=e.exec(t);)r.push(n);return r}if(e.global){let t=r(e,this.original);t.forEach(e=>{if(e.index!=null){let t=n(e,this.original);t!==e[0]&&this.overwrite(e.index,e.index+e[0].length,t)}})}else{let t=this.original.match(e);if(t&&t.index!=null){let e=n(t,this.original);e!==t[0]&&this.overwrite(t.index,t.index+t[0].length,e)}}return this}_replaceString(e,t){let{original:n}=this,r=n.indexOf(e);return r!==-1&&(typeof t==`function`&&(t=t(e,r,n)),e!==t&&this.overwrite(r,r+e.length,t)),this}replace(e,t){return typeof e==`string`?this._replaceString(e,t):this._replaceRegexp(e,t)}_replaceAllString(e,t){let{original:n}=this,r=e.length;for(let i=n.indexOf(e);i!==-1;i=n.indexOf(e,i+r)){let e=n.slice(i,i+r),a=t;typeof t==`function`&&(a=t(e,i,n)),e!==a&&this.overwrite(i,i+r,a)}return this}replaceAll(e,t){if(typeof e==`string`)return this._replaceAllString(e,t);if(!e.global)throw TypeError(`MagicString.prototype.replaceAll called with a non-global RegExp argument`);return this._replaceRegexp(e,t)}};const N=n.default||n;function P(){let t=`virtual:404-page`,n=`import React from "react";
import { createClient } from "@metagptx/web-sdk";
import { useMemo, useState } from "react";

const client = createClient();

const Icon = ({ vb, s = 18, fill = "none", stroke = "currentColor", sw = 2, children }: any) => (
  <svg width={s} height={s} viewBox={vb} fill={fill} stroke={stroke} strokeWidth={sw} strokeLinecap="round" strokeLinejoin="round">
    {children}
  </svg>
);

const styles = {
  container: { minHeight: "100vh", display: "flex", flexDirection: "column", alignItems: "center", justifyContent: "center", background: "#f6f6f6", padding: 24, textAlign: "center" as const },
  content: { display: "flex", flexDirection: "column", alignItems: "center", gap: 24, maxWidth: 420 },
  errorCodeWrapper: { display: "flex", alignItems: "center", justifyContent: "center" },
  errorCode: { fontSize: 120, fontWeight: 800, lineHeight: 1, color: "#4267ff", margin: 0, letterSpacing: -4, textShadow: "0 4px 20px rgba(66, 103, 255, 0.3)" },
  textGroup: { display: "flex", flexDirection: "column", alignItems: "center", gap: 8 },
  titleRow: { display: "flex", alignItems: "center", gap: 8, margin: 0 },
  pathText: { fontSize: 20, fontWeight: 600, color: "#4267ff" },
  titleText: { fontSize: 20, fontWeight: 600, color: "rgba(12, 12, 12, 0.95)" },
  description: { fontSize: 15, lineHeight: 1.5, color: "rgba(12, 12, 12, 0.55)", margin: 0 },
  buttonGroup: { display: "flex", gap: 12, justifyContent: "center", flexWrap: "wrap" as const },
  btnBase: { display: "inline-flex", alignItems: "center", justifyContent: "center", gap: 8, height: 44, padding: "0 24px", fontSize: 15, fontWeight: 500, borderRadius: 22, cursor: "pointer", transition: "all 0.2s ease" },
  outline: { color: "#4267ff", background: "transparent", border: "1.5px solid #4267ff" },
  outlineHover: { background: "rgba(66, 103, 255, 0.08)" },
  primary: { color: "#fff", background: "#4267ff", border: "none" },
  primaryHover: { background: "#425ce1" },

  overlay: { position: "fixed" as const, inset: 0, background: "rgba(12, 12, 12, 0.4)", display: "flex", alignItems: "center", justifyContent: "center", zIndex: 1000 },
  modal: { background: "#fff", borderRadius: 16, width: "90%", maxWidth: 480, boxShadow: "0 20px 60px rgba(12, 12, 12, 0.2)", overflow: "hidden" },
  modalHeader: { display: "flex", alignItems: "center", justifyContent: "space-between", padding: "20px 24px", borderBottom: "1px solid rgba(12, 12, 12, 0.08)" },
  modalTitle: { fontSize: 18, fontWeight: 600, color: "rgba(12, 12, 12, 0.95)", margin: 0 },
  closeButton: { display: "flex", alignItems: "center", justifyContent: "center", width: 32, height: 32, background: "transparent", border: "none", borderRadius: 8, cursor: "pointer", color: "rgba(12, 12, 12, 0.55)", transition: "all 0.2s ease" },
  closeButtonHover: { background: "rgba(12, 12, 12, 0.06)", color: "rgba(12, 12, 12, 0.95)" },
  modalBody: { padding: "20px 24px" },
  textarea: { width: "100%", minHeight: 140, padding: "14px 16px", fontSize: 15, lineHeight: 1.5, color: "rgba(12, 12, 12, 0.95)", background: "#fff", border: "1.5px solid rgba(12, 12, 12, 0.12)", borderRadius: 12, resize: "vertical" as const, outline: "none", fontFamily: "inherit", boxSizing: "border-box" as const, transition: "border-color 0.2s ease" },
  textareaFocus: { borderColor: "#4267ff" },
  modalFooter: { display: "flex", alignItems: "center", justifyContent: "center", gap: 12, padding: "16px 24px 24px" },
};

export default function NotFoundPage() {
  const isInMGXIframe = window.self !== window.top && window.name?.includes("devIframe");
  const currentPath = window.location.pathname;

  const [showModal, setShowModal] = useState(false);
  const [context, setContext] = useState("");
  const [hover, setHover] = useState({ home: false, primary: false, close: false, cancel: false, submit: false, focus: false });

  const S = useMemo(
    () => ({
      outline: { ...styles.btnBase, ...styles.outline },
      primary: { ...styles.btnBase, ...styles.primary },
    }),
    []
  );

  const h = (k: keyof typeof hover, v: boolean) => setHover((s) => ({ ...s, [k]: v }));
  const create = () => (client.frame.createPage({ prompt: context || undefined }), setShowModal(false), setContext(""));

  const HomeIcon = (
    <Icon vb="0 0 24 24">
      <path d="M3 9l9-7 9 7v11a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2z" />
      <polyline points="9 22 9 12 15 12 15 22" />
    </Icon>
  );
  const PlusIcon = (
    <Icon vb="0 0 24 24">
      <rect x="3" y="3" width="18" height="18" rx="2" ry="2" />
      <line x1="12" y1="8" x2="12" y2="16" />
      <line x1="8" y1="12" x2="16" y2="12" />
    </Icon>
  );
  const CloseIcon = (
    <Icon vb="0 0 24 24" s={20}>
      <line x1="18" y1="6" x2="6" y2="18" />
      <line x1="6" y1="6" x2="18" y2="18" />
    </Icon>
  );
  const AtomsIcon = (
    <Icon vb="0 0 1024 1024" fill="currentColor" stroke="none" sw={0}>
      <path d="M274.56 842.88c3.584-51.456 100.032-86.656 215.36-78.528 115.264 8.064 205.824 56.32 202.24 107.84-3.584 51.456-100.032 86.656-215.296 78.592-115.328-8.128-205.888-56.384-202.24-107.904zM700.672 547.712c27.072-119.616 105.536-203.776 175.168-187.968 69.632 15.808 104.128 125.632 76.992 245.248-27.136 119.616-105.6 203.712-175.168 187.968-69.632-15.808-104.128-125.632-76.992-245.248zM110.08 308.352C192.448 140.16 351.296 48.832 465.024 104.448c113.728 55.68 139.264 237.184 56.96 405.44-82.176 168.256-241.088 259.52-354.816 203.904-113.728-55.68-139.264-237.184-56.96-405.44zM589.248 83.2c17.6-26.24 95.808-4.608 174.72 48.256 78.848 52.864 128.64 116.992 111.04 143.232-17.536 26.24-95.744 4.672-174.72-48.192C621.44 173.632 571.776 109.44 589.312 83.2z" />
    </Icon>
  );

  return (
    <div style={styles.container}>
      <div style={styles.content}>
        <div style={styles.errorCodeWrapper}>
          <h1 style={styles.errorCode}>404</h1>
        </div>

        <div style={styles.textGroup}>
          <div style={styles.titleRow}>
            <span style={styles.pathText}>{currentPath}</span>
            <span style={styles.titleText}>not found</span>
          </div>
          <p style={styles.description}>This page is not yet created.</p>
        </div>

        <div style={styles.buttonGroup}>
          <a
            href="/"
            style={{ ...S.outline, ...(hover.home ? styles.outlineHover : {}), textDecoration: "none" }}
            onMouseEnter={() => h("home", true)}
            onMouseLeave={() => h("home", false)}
          >
            {HomeIcon} Go Home
          </a>

          {isInMGXIframe ? (
            <button
              style={{ ...S.primary, ...(hover.primary ? styles.primaryHover : {}) }}
              onMouseEnter={() => h("primary", true)}
              onMouseLeave={() => h("primary", false)}
              onClick={() => setShowModal(true)}
            >
              {PlusIcon} Create
            </button>
          ) : (
            <button
              style={{ ...S.primary, ...(hover.primary ? styles.primaryHover : {}) }}
              onMouseEnter={() => h("primary", true)}
              onMouseLeave={() => h("primary", false)}
              onClick={() => (window.location.href = "https://atoms.dev")}
            >
              {AtomsIcon} Go to Atoms
            </button>
          )}
        </div>
      </div>

      {showModal && (
        <div style={styles.overlay} onClick={() => setShowModal(false)}>
          <div style={styles.modal} onClick={(e) => e.stopPropagation()}>
            <div style={styles.modalHeader}>
              <h3 style={styles.modalTitle}>Create Page</h3>
              <button
                style={{ ...styles.closeButton, ...(hover.close ? styles.closeButtonHover : {}) }}
                onMouseEnter={() => h("close", true)}
                onMouseLeave={() => h("close", false)}
                onClick={() => setShowModal(false)}
              >
                {CloseIcon}
              </button>
            </div>

            <div style={styles.modalBody}>
              <textarea
                style={{ ...styles.textarea, ...(hover.focus ? styles.textareaFocus : {}) }}
                placeholder="Add any additional context or requirements for the page..."
                value={context}
                onChange={(e) => setContext(e.target.value)}
                onFocus={() => h("focus", true)}
                onBlur={() => h("focus", false)}
              />
            </div>

            <div style={styles.modalFooter}>
              <button
                style={{ ...S.outline, ...(hover.cancel ? styles.outlineHover : {}) }}
                onMouseEnter={() => h("cancel", true)}
                onMouseLeave={() => h("cancel", false)}
                onClick={() => setShowModal(false)}
              >
                Cancel
              </button>
              <button
                style={{ ...S.primary, ...(hover.submit ? styles.primaryHover : {}) }}
                onMouseEnter={() => h("submit", true)}
                onMouseLeave={() => h("submit", false)}
                onClick={create}
              >
                {PlusIcon} Create Page
              </button>
            </div>
          </div>
        </div>
      )}
    </div>
  );
}
`;function i(e){for(let t of e)if(r.isJSXAttribute(t)&&r.isJSXIdentifier(t.name,{name:`path`})&&r.isStringLiteral(t.value))return t.value.value}function o(e,t){return e.source.value.includes(t)}function s(e,t){return e.specifiers.some(e=>r.isImportDefaultSpecifier(e)||r.isImportSpecifier(e)?e.local.name===t:!1)}return{name:`vite-plugin-404`,enforce:`pre`,resolveId(e){return e===t||e.startsWith(t)?`\0${t}.tsx`:null},async load(e){if(e===`\0${t}.tsx`){let t=await a(n,e,{loader:`tsx`});return{code:t.code,map:t.map}}return null},transform(n,a){if(a===`\0${t}.tsx`||!a.includes(`App.tsx`)||!a.endsWith(`.tsx`))return null;let c;try{c=e.parse(n,{sourceType:`module`,plugins:[`jsx`,`typescript`]})}catch{return null}let l=!1,u=!1,d=!1,f=-1,p=-1,m=`      `;N(c,{ImportDeclaration(e){let t=e.node;o(t,`virtual:404-page`)&&(l=!0),s(t,`NotFoundPage`)&&(u=!0),t.end&&t.end>f&&(f=t.end)},JSXElement(e){let t=e.node.openingElement;if(r.isJSXIdentifier(t.name,{name:`Route`})){let e=i(t.attributes);if(e===`*`&&(d=!0),t.start!==void 0&&t.start!==null){let e=n.lastIndexOf(`
`,t.start)+1,r=n.substring(e,t.start);/^\s+$/.test(r)&&(m=r)}}if(r.isJSXIdentifier(t.name,{name:`Routes`})){let t=e.node.closingElement;t&&t.start!==void 0&&t.start!==null&&(p=t.start)}}});let h=!l&&!u,g=!d;if(!h&&!g||h&&f===-1)return null;let _=new M(n),v=!1;if(h&&f!==-1){let e=`
import NotFoundPage from 'virtual:404-page';`;_.appendRight(f,e),v=!0}if(g&&p!==-1){let e=`${m}<Route path="*" element={<NotFoundPage />} />\n`;_.appendLeft(p,e),v=!0}return v?{code:_.toString(),map:_.generateMap({hires:!0})}:null}}}const F=n.default||n;function I(e={}){let{routeFile:n=`src/App.tsx`,entryFile:r=`src/main.tsx`,moduleId:a=`virtual:routes-manifest`,notifyParent:c=!0,messageType:l=`routes-manifest`,excludePaths:u=[`*`]}=e,d=`\0${a}`,f=process.cwd();function p(e,t){for(let n of e)if(i.isJSXAttribute(n)&&i.isJSXIdentifier(n.name,{name:t})&&i.isStringLiteral(n.value))return n.value.value}function m(e){for(let t of e)if(i.isJSXAttribute(t)&&i.isJSXIdentifier(t.name,{name:`element`})){let e=t.value;if(i.isJSXExpressionContainer(e)){let t=e.expression;if(i.isJSXElement(t)){let e=t.openingElement;if(i.isJSXIdentifier(e.name))return e.name.name}}}}function h(e){return e.some(e=>i.isJSXAttribute(e)&&i.isJSXIdentifier(e.name,{name:`element`}))}function g(e){return e.some(e=>i.isJSXAttribute(e)&&i.isJSXIdentifier(e.name,{name:`index`}))}function _(e,t){let n=[];for(let r of e)if(i.isJSXElement(r)){let e=r.openingElement;if(i.isJSXIdentifier(e.name,{name:`Route`})){let e=v(r,t);e&&n.push(e)}}return n}function v(e,t){let n=e.openingElement.attributes,r=p(n,`path`),i=g(n),a=h(n),o=m(n),s;if(i)s=t||`/`;else if(r===void 0)return null;else s=r.startsWith(`/`)?r:t===`/`?`/${r}`:`${t}/${r}`;s=s.replace(/\/+/g,`/`),s!==`/`&&s.endsWith(`/`)&&(s=s.slice(0,-1));let c={path:s,hasElement:a,elementName:o,isWildcard:r===`*`,isIndex:i};if(e.children&&e.children.length>0){let t=_(e.children,s);t.length>0&&(c.children=t)}return c}function y(e,t=[]){for(let n of e)n.hasElement&&t.push({path:n.path,elementName:n.elementName}),n.children&&y(n.children,t);return t}function b(e){let t=0;for(let n of e)t++,n.children&&(t+=b(n.children));return t}function x(e){let n={routeTree:[],routes:[],pageCount:0,total:0,generatedAt:new Date().toISOString()};if(!o.existsSync(e))return console.warn(`[routes-scanner] Route file not found: ${e}`),n;let r=o.readFileSync(e,`utf-8`),a;try{a=t.parse(r,{sourceType:`module`,plugins:[`jsx`,`typescript`]})}catch(e){return console.error(`[routes-scanner] Failed to parse route file:`,e),n}let s=[];F(a,{JSXElement(e){let t=e.node.openingElement;if(i.isJSXIdentifier(t.name,{name:`Routes`})){let t=e.node.children;for(let e of t)if(i.isJSXElement(e)){let t=e.openingElement;if(i.isJSXIdentifier(t.name,{name:`Route`})){let t=v(e,``);t&&s.push(t)}}e.stop()}}});let c=y(s),l=c.filter(e=>!u.some(t=>e.path===t||e.path.match(RegExp(`^${t.replace(`*`,`.*`)}$`))));return{routeTree:s,routes:c,pageCount:l.length,total:b(s),generatedAt:new Date().toISOString()}}return{name:`vite-plugin-routes-scanner`,configResolved(e){f=e.root},resolveId(e){if(e===a)return d},load(e){if(e===d){let e=s.resolve(f,n),t=x(e),r=c?`
// 向父级 iframe 发送路由信息
function notifyParentWindow() {
  console.log('[routes-scanner] notifyParentWindow', routesManifest);
  if (typeof window !== 'undefined' && window.parent && window.parent !== window) {
    try {
      window.parent.postMessage({
        type: '${l}',
        targetName: window.name,
        data: routesManifest,
      }, '*');
    } catch (e) {
      console.warn('[routes-scanner] Failed to notify parent window:', e);
    }
  }
}

// 页面加载完成后发送
if (typeof window !== 'undefined') {
  if (document.readyState === 'complete') {
    notifyParentWindow();
  } else {
    window.addEventListener('load', notifyParentWindow);
  }
}

export { notifyParentWindow };
`:``;return`
export const routesManifest = ${JSON.stringify(t,null,2)};
export const routeTree = routesManifest.routeTree;
export const routes = routesManifest.routes;
export const pageCount = routesManifest.pageCount;
export const total = routesManifest.total;
${r}
export default routesManifest;
`}},transform(e,t){let n=t.replace(/\\/g,`/`),i=s.resolve(f,r).replace(/\\/g,`/`);if(n===i&&c){let t=`import "${a}";\n`;return e.includes(a)?null:{code:t+e,map:null}}return null},configureServer(e){let t=s.resolve(f,n);e.watcher.add(t),e.watcher.on(`change`,n=>{if(n===t){let t=e.moduleGraph.getModuleById(d);t&&(e.moduleGraph.invalidateModule(t),e.ws.send({type:`full-reload`}))}})}}}function L(){return`
import { createHotContext } from '/@vite/client';

const hot = createHotContext('/__mgx_vite_error_overlay__.js');

if (window.__MGX_VITE_ERROR_OVERLAY_INSTALLED__) {
  hot.on('vite:beforeUpdate', function() {
    const overlay = document.getElementById('custom-vite-error-overlay');
    if (overlay) overlay.remove();
  });
} else {
  window.__MGX_VITE_ERROR_OVERLAY_INSTALLED__ = true;

let overlayElement = null;
let cleanupRegistered = false;
let latestOverlayError = null;
let chatInfoStatus = 'not_working';
const mascotImage = 'https://public-frontend-cos.metadl.com/commonfile/engineer.svg';
const isInMGXIframe = window.self !== window.top && window.name && window.name.includes('devIframe');

function escapeHtml(str) {
  if (!str) return '';
  return str
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#39;');
}

function removeOverlay() {
  if (overlayElement) {
    overlayElement.remove();
    overlayElement = null;
  }

  if (cleanupRegistered) {
    document.removeEventListener('keydown', onEscape);
    cleanupRegistered = false;
  }
}

function getOverlayTitle() {
  if (chatInfoStatus === 'working') {
    return 'Agent is on the job! Please check back in a bit.';
  }

  return 'Something is wrong';
}

function onEscape(e) {
  if (e.key === 'Escape') removeOverlay();
}

function getErrorLocation(err) {
  if (!err) {
    return '';
  }

  const file = err.id || (err.loc && err.loc.file) || '';
  const line = err.loc && typeof err.loc.line === 'number' ? err.loc.line : null;
  const column = err.loc && typeof err.loc.column === 'number' ? err.loc.column : null;

  if (!file) {
    return '';
  }

  if (line === null) {
    return file;
  }

  if (column === null) {
    return file + ':' + line;
  }

  return file + ':' + line + ':' + column;
}

function getPrimaryMessage(err) {
  const message = err && err.message ? String(err.message) : 'Unknown Vite error';
  const firstLine = message.split(/\\r?\\n/).find(Boolean);

  return firstLine || 'Unknown Vite error';
}

function buildErrorDetails(err) {
  const parts = [];
  const message = err && err.message ? String(err.message).trim() : '';
  const frame = err && err.frame ? String(err.frame).trim() : '';

  if (message) {
    parts.push(message);
  }

  if (frame && (!message || !message.includes(frame))) {
    parts.push(frame);
  }

  return parts.filter(Boolean).join('\\n\\n');
}

function normalizeViteError(payload) {
  const source = payload && (payload.err || payload);

  if (!source) {
    return {
      message: 'Unknown Vite error',
      frame: '',
      stack: '',
      id: '',
      loc: null,
      plugin: '',
      details: 'Unknown Vite error',
    };
  }

  if (typeof source === 'string') {
    return {
      message: source,
      frame: '',
      stack: '',
      id: '',
      loc: null,
      plugin: '',
      details: source,
    };
  }

  return {
    message: source.message || 'Unknown Vite error',
    frame: source.frame || '',
    stack: source.stack || '',
    id: source.id || '',
    loc: source.loc || null,
    plugin: source.plugin || '',
    details: buildErrorDetails(source),
  };
}

function postResolveMessage(err) {
  const message = {
    type: 'errorResolve',
    targetName: window.name || '',
    data: {
      message: err.message || 'Unknown error',
      stack: err.stack || '',
      frame: err.frame || '',
      id: err.id || '',
      loc: err.loc || null,
      plugin: err.plugin || '',
      details: err.details || buildErrorDetails(err),
      errorType: 'compile',
    },
  };

  try {
    if (window.top) {
      window.top.postMessage(message, '*');
    }

    if (window.opener && window.opener !== window.top) {
      window.opener.postMessage(message, '*');
    }
  } catch (error) {
    console.warn('Failed to send resolve message:', error);
  }
}

function requestChatInfo() {
  if (!isInMGXIframe) {
    return;
  }

  const message = {
    type: 'chat_info',
    targetName: window.name || '',
  };

  try {
    if (window.top) {
      window.top.postMessage(message, '*');
    }

    if (window.opener && window.opener !== window.top) {
      window.opener.postMessage(message, '*');
    }
  } catch (error) {
    console.warn('Failed to request chat_info:', error);
  }
}

function createOverlay(err, shouldRequestChatInfo = true) {
  removeOverlay();
  latestOverlayError = err;

  if (shouldRequestChatInfo) {
    requestChatInfo();
  }

  const overlay = document.createElement('div');
  overlay.id = 'custom-vite-error-overlay';

  overlay.innerHTML =
    '<div style="position:fixed;inset:0;z-index:99999;background:rgba(255,255,255,0.92);backdrop-filter:blur(10px);display:flex;align-items:center;justify-content:center;padding:24px;font-family:IBM Plex Sans,-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif;">' +
      '<div style="position:relative;width:min(920px,100%);min-height:min(80vh,760px);display:flex;align-items:center;justify-content:center;padding:24px;">' +
        '<div style="position:relative;display:flex;flex-direction:column;align-items:center;justify-content:center;gap:8px;width:301px;max-width:100%;">' +
          '<div style="width:110px;height:77.536px;opacity:0.8;overflow:hidden;display:flex;align-items:center;justify-content:center;">' +
            '<img alt="" src="' + mascotImage + '" style="width:80px;height:80px;object-fit:cover;pointer-events:none;" />' +
          '</div>' +
         
          '<div style="font-size:14px;line-height:24px;font-weight:400;color:rgba(12,12,12,0.8);text-align:center;">' + getOverlayTitle() + '</div>' +
           
          '<div style="display:flex;align-items:flex-start;gap:8px;">' +
          '<button id="cve-reload" style="display:inline-flex;align-items:center;justify-content:center;gap:4px;height:28px;padding:0 12px;border:none;border-radius:8px;background:rgba(12,12,12,0.06);color:rgba(12,12,12,0.95);font-size:14px;font-weight:500;line-height:22px;cursor:pointer;">' +
              '<svg xmlns="http://www.w3.org/2000/svg" width="12" height="12" viewBox="0 0 12 12" fill="none"><path d="M11.105 5.20813C11.3465 5.20827 11.5425 5.40409 11.5425 5.64563C11.5425 8.78666 8.99598 11.333 5.85498 11.3331C4.11675 11.3331 2.56217 10.5521 1.51929 9.32422V10.1779C1.51929 10.4195 1.32341 10.6154 1.08179 10.6154C0.840163 10.6154 0.644287 10.4195 0.644287 10.1779V8.26892C0.644287 8.0273 0.840163 7.83142 1.08179 7.83142H2.99072C3.23235 7.83142 3.42822 8.0273 3.42822 8.26892C3.42814 8.51048 3.2323 8.70642 2.99072 8.70642H2.14392C3.02711 9.77583 4.36095 10.4581 5.85498 10.4581C8.51273 10.458 10.6675 8.30342 10.6675 5.64563C10.6675 5.40401 10.8634 5.20813 11.105 5.20813Z" fill="#0C0C0C" fill-opacity="0.95"/><path d="M5.6875 0C7.42573 0 8.98031 0.781055 10.0232 2.00891V1.15527C10.0232 0.913649 10.2191 0.717773 10.4607 0.717773C10.7023 0.717773 10.8982 0.913649 10.8982 1.15527V3.06421C10.898 3.30566 10.7022 3.50171 10.4607 3.50171H8.55176C8.31026 3.50171 8.11446 3.30566 8.11426 3.06421C8.11434 2.82265 8.31018 2.62671 8.55176 2.62671H9.39856C8.51537 1.5573 7.18153 0.875 5.6875 0.875C3.02975 0.87514 0.875 3.02971 0.875 5.6875C0.874798 5.92895 0.679 6.125 0.4375 6.125C0.196119 6.12486 0.000202109 5.92887 0 5.6875C0 2.54647 2.5465 0.000140261 5.6875 0Z" fill="#0C0C0C" fill-opacity="0.95"/></svg>' +
              '<span>Reload</span>' +
            '</button>' +
            (isInMGXIframe && chatInfoStatus !== 'working'
              ? '<button id="cve-resolve" style="display:inline-flex;align-items:center;justify-content:center;gap:4px;height:28px;padding:0 12px;border:none;border-radius:8px;background:#111111;color:rgba(255,255,255,0.95);font-size:14px;font-weight:500;line-height:22px;cursor:pointer;">' +
                  '<svg xmlns="http://www.w3.org/2000/svg" width="14" height="14" viewBox="0 0 14 14" fill="none"><path d="M12.4186 3.77422C12.3921 3.70862 12.35 3.65045 12.296 3.60467C12.2421 3.5589 12.1778 3.5269 12.1088 3.51141C12.0397 3.49593 11.968 3.49742 11.8996 3.51576C11.8313 3.5341 11.7684 3.56874 11.7164 3.61672L9.5125 5.65054L8.57023 5.4482L8.36789 4.50593L10.4017 2.30203C10.4497 2.25002 10.4843 2.18714 10.5027 2.1188C10.521 2.05046 10.5225 1.97869 10.507 1.90965C10.4915 1.84061 10.4595 1.77635 10.4138 1.72239C10.368 1.66843 10.3098 1.62638 10.2442 1.59984C9.64645 1.35803 8.99844 1.26673 8.35713 1.33396C7.71581 1.40119 7.10083 1.62489 6.5662 1.98542C6.03158 2.34594 5.59367 2.83224 5.29097 3.4016C4.98827 3.97096 4.83003 4.60595 4.83015 5.25078C4.82945 5.79529 4.94115 6.33409 5.15828 6.83343L1.86554 9.68047C1.85734 9.68703 1.84968 9.69468 1.84203 9.70179C1.5138 10.03 1.32941 10.4752 1.32941 10.9394C1.32941 11.1692 1.37468 11.3968 1.46263 11.6091C1.55059 11.8215 1.67951 12.0144 1.84203 12.177C2.00455 12.3395 2.19749 12.4684 2.40983 12.5563C2.62218 12.6443 2.84977 12.6896 3.07961 12.6896C3.54379 12.6896 3.98896 12.5052 4.31718 12.177C4.32429 12.1698 4.33195 12.1616 4.33851 12.154L7.185 8.86015C7.78461 9.12336 8.4404 9.23287 9.09299 9.17876C9.74559 9.12466 10.3744 8.90865 10.9225 8.55028C11.4705 8.19192 11.9206 7.70253 12.2319 7.1264C12.5431 6.55028 12.7058 5.90562 12.7052 5.25078C12.706 4.74469 12.6087 4.24324 12.4186 3.77422ZM8.76765 8.31328C8.2498 8.31257 7.74054 8.18088 7.28726 7.93047C7.19945 7.88195 7.09762 7.86509 6.99886 7.88272C6.90009 7.90034 6.81038 7.95137 6.74476 8.02726L3.68773 11.5677C3.5223 11.7249 3.30202 11.8112 3.07385 11.8083C2.84569 11.8054 2.62769 11.7134 2.46634 11.5521C2.30499 11.3907 2.21305 11.1727 2.21013 10.9446C2.20721 10.7164 2.29353 10.4961 2.4507 10.3307L5.98843 7.27422C6.06446 7.20857 6.11557 7.11876 6.1332 7.01987C6.15082 6.92098 6.13389 6.81904 6.08523 6.73117C5.80634 6.22675 5.67546 5.65387 5.70763 5.07839C5.7398 4.50291 5.93372 3.94819 6.2671 3.478C6.60047 3.00782 7.05977 2.64125 7.59218 2.42045C8.12459 2.19965 8.70851 2.13359 9.27679 2.22984L7.57054 4.07883C7.52307 4.13033 7.48865 4.19247 7.47019 4.26004C7.45173 4.32761 7.44976 4.39862 7.46445 4.46711L7.77398 5.90703C7.79169 5.98938 7.83279 6.06488 7.89235 6.12444C7.95192 6.184 8.02741 6.22511 8.10976 6.24281L9.55078 6.55234C9.61927 6.56703 9.69028 6.56506 9.75784 6.5466C9.82541 6.52814 9.88756 6.49372 9.93906 6.44625L11.788 4.74C11.8618 5.17914 11.8389 5.62907 11.7211 6.0585C11.6033 6.48792 11.3934 6.88654 11.106 7.22661C10.8185 7.56669 10.4604 7.84006 10.0566 8.02772C9.65279 8.21538 9.21294 8.31283 8.76765 8.31328Z" fill="white" fill-opacity="0.95"/></svg>' +
                  '<span>Resolve</span>' +
                '</button>'
              : '') +
          '</div>' +
          
        '</div>' +
      '</div>' +
    '</div>';

  const backdrop = overlay.firstElementChild;
  const card = backdrop && backdrop.firstElementChild;

  if (card) {
    card.addEventListener('click', function(e) {
      e.stopPropagation();
    });
  }

  const reloadButton = overlay.querySelector('#cve-reload');
  if (reloadButton) {
    reloadButton.addEventListener('click', function() {
      window.location.reload();
    });
  }

  const resolveButton = overlay.querySelector('#cve-resolve');
  if (resolveButton) {
    resolveButton.addEventListener('click', function(e) {
      e.stopPropagation();
      postResolveMessage(err);
    });
  }

  document.addEventListener('keydown', onEscape);
  cleanupRegistered = true;
  document.body.appendChild(overlay);
  overlayElement = overlay;
}

window.addEventListener('message', function(event) {
  const payload = event.data;
  if (!payload || payload.type !== 'chat_info' || !payload.data) {
    return;
  }

  const nextStatus = payload.data.status;
  if (nextStatus !== 'working' && nextStatus !== 'not_working') {
    return;
  }

  chatInfoStatus = nextStatus;

  if (overlayElement && latestOverlayError) {
    createOverlay(latestOverlayError, false);
  }
});

hot.on('vite:error', function(data) {
  createOverlay(normalizeViteError(data));
});

hot.on('vite:beforeUpdate', function() {
  removeOverlay();
});
}
`}function R(){return{name:`vite-plugin-error`,apply:`serve`,config(){return{server:{hmr:{overlay:!1}}}},transformIndexHtml(){return[{tag:`script`,attrs:{type:`module`},children:L(),injectTo:`head-prepend`}]}}}function z(e={}){let{404:t={},error:n={},routes:r={}}=e,{enable:i=!0}=t,{enable:a=!0}=n,{enable:o=!0,...s}=r,c=[];return a&&c.push(R()),i&&c.push(P()),o&&c.push(I(s)),c}export{z as atoms,R as vitePluginError};
